gateway/ln-gateway/src/bin/cln_extension.rs

```diff
@@ -295,24 +304,100 @@ pub enum ClnRpcError {
+            let payment_hash = payload.htlc.payment_hash.to_vec();
+
+            match subscription
+                .send(Ok(SubscribeInterceptHtlcsResponse {
+                    payment_hash: payment_hash.clone(),
+                    incoming_amount_msat: payload.htlc.amount.msats,
+                    outgoing_amount_msat: payload.onion.forward_amount.msats,
+                    incoming_expiry: htlc_expiry,
+                    short_channel_id,
+                    // TODO: generate unique id for each intercepted HTLC
+                    intercepted_htlc_id: payment_hash.clone(),
+                }))
+                .await
+            {
+                Ok(_) => {
+                    // Open a channel to receive the outcome of the HTLC processing
+                    let (sender, receiver) = oneshot::channel::<Action>();
+                    self.outcomes.lock().await.insert(short_channel_id, sender);
+
+                    // If the gateway does not respond within the HTLC expiry,
+                    // Automatically respond with a failure message.
+                    return tokio::time::timeout(
+                        Duration::from_secs(5),
+                        self.await_htlc_processing(receiver, payment_hash.clone()),
+                    )
+                    .await
+                    .unwrap_or_else(|e| {
+                        error!("await_htlc_processing error {:?}", e);
+
+                        // Bolt 4: https://github.com/lightning/bolts/blob/master/04-onion-routing.md#failure-messages
+                        // 2002 error code reports general temporary failure of this processing node.
+                        serde_json::json!({
+                            "result": "fail",
+                            "failure_message": "2002"
+                        })
+                    });
+                }
+                Err(e) => {
+                    error!("Failed to send htlc to subscription: {:?}", e);
+
+                    // BOLT 4: https://github.com/lightning/bolts/blob/master/04-onion-routing.md#failure-messages
+                    // 2002 error code reports general temporary failure of this processing node.
+                    return serde_json::json!({
+                        "result": "fail",
+                        "failure_message": "2002"
+                    });
+                }
+            }
         }
+
+        // We have no subscription for this HTLC.
+        // Ignore it by requesting the node to continue
+        serde_json::json!({ "result": "continue" })
     }
 
-    async fn intercept_htlc(&self, _payload: HtlcAccepted) -> serde_json::Value {
-        unimplemented!("TODO: Implement HTLC filtering using short channel ids")
+    // Await outcome of HTLC processing by the gateway.
+    // If the gateway does not respond within the HTLC expiry `cltv_expiry_relative`,
+    // Automatically respond with a failure message.
+    async fn await_htlc_processing(
+        &self,
```

Comment:
Docstrings should begin with `///` so they are rendered on https://docs.fedimint.org